File: _baseline/swarm.py
#!/usr/bin/env python3
"""ModelSwarm — the real models as agents on the stigmergent field (no central orchestrator).

Each model is an agent. It does not message the others. It SENSES the field's current leaning,
forms its own pick, and DEPOSITS a recruit-signal on that pick. The field — with its decay and
quorum — is the entire coordination medium. Consensus emerges from traces, not from a tally
function. That is the difference between this and majority voting: later agents are influenced
by the trail earlier agents left (true stigmergy), and unreinforced options fade between rounds.

Agents are (backend, model) specs — the same swappable organs the rotator uses. Pass 2 or 200;
the mechanism doesn't change. Heterogeneous models (different vendors, sizes) coordinate through
one shared field.

Decision = the option carrying the strongest live recruit-signal, gated by quorum (enough
DISTINCT models behind it). If quorum is unmet, the swarm abstains rather than forcing a call —
the same abstention discipline as the belief layer.
"""
from __future__ import annotations
import re
from typing import List, Tuple, Optional, Dict, Sequence

from stigmergy import StigmergicField

KIND = "recruit"

def _norm(label: str) -> str:
    return re.sub(r"[^a-z0-9]+", "_", label.strip().lower()).strip("_")

def _match_option(text: str, options: Sequence[str]) -> Optional[str]:
    """Map a model's free-text reply to one option label, or None (abstain) if ambiguous."""
    t = (text or "").lower()
    hits = [o for o in options if o.lower() in t]
    if len(hits) == 1:
        return hits[0]
    # fall back to first-line exact-ish token match
    first = t.strip().splitlines()[0] if t.strip() else ""
    for o in options:
        if _norm(o) == _norm(first) or o.lower() == first.strip(" .:-"):
            return o
    return hits[0] if len(hits) >= 1 else None


class ModelSwarm:
    def __init__(self, specs: List[Tuple], field: Optional[StigmergicField] = None):
        if not specs:
            raise ValueError("ModelSwarm needs at least one (backend, model) spec")
        self.specs = list(specs)
        self.field = field or StigmergicField()

    def _ask_agent(self, backend, model: str, prompt: str, options: Sequence[str],
                   leader: Optional[str]) -> Optional[str]:
        """One agent's turn: see the field's current leaning, then pick one option."""
        opts = " | ".join(options)
        hint = f"\nThe swarm is currently leaning toward: {leader}." if leader else ""
        sys_msg = ("You are one agent in a swarm deciding a single question. Weigh the swarm's "
                   "current leaning, but choose what you actually judge correct. Reply with ONLY "
                   "one option label, nothing else.")
        user = f"Question: {prompt}\nOptions: {opts}.{hint}\nYour one-word choice:"
        msgs = [{"role": "system", "content": sys_msg}, {"role": "user", "content": user}]
        try:
            out = backend.chat(msgs, model=model, options={"temperature": 0.3, "num_predict": 12})
        except Exception:
            return None
        return _match_option(out, options)

    def coordinate(self, prompt: str, options: Sequence[str], rounds: int = 2,
                   quorum_min: int = 0) -> Dict:
        """Run the swarm. Returns picks, field snapshot, decision, quorum status."""
        if quorum_min <= 0:
            quorum_min = max(2, (len(self.specs) // 2) + 1)   # simple majority of agents
        history: List[Dict] = []
        for _ in range(max(1, rounds)):
            # current leader from the field (what earlier traces say)
            sensed = self.field.sense_all(KIND) if hasattr(self.field, "sense_all") else {}
            leader = max(sensed, key=sensed.get) if sensed else None
            round_picks = {}
            for backend, model in self.specs:
                pick = self._ask_agent(backend, model, prompt, options, leader)
                round_picks[model] = pick
                if pick:
                    self.field.deposit(KIND, _norm(pick), strength=0.34, agent=model)
            history.append(round_picks)
        # decision = strongest live recruit signal across option topics
        scores = {o: self.field.sense(_norm(o), kinds=[KIND]).get(KIND, 0.0) for o in options}
        decision = max(scores, key=scores.get) if any(scores.values()) else None
        quorum_met = bool(decision) and self.field.quorum(KIND, _norm(decision),
                                                          min_depositors=quorum_min, min_strength=0.0)
        return {"decision": decision if quorum_met else None,
                "leader_raw": decision, "quorum_met": quorum_met, "quorum_min": quorum_min,
                "scores": {o: round(s, 4) for o, s in scores.items()},
                "rounds": history, "n_agents": len(self.specs)}


# ================================================================ self-test
def _stub_test():
    """Deterministic stub models prove the coordination mechanism without a network."""
    class Stub:
        def __init__(self, answer): self.answer = answer
        def chat(self, messages, model=None, options=None): return self.answer
    # 5 agents: 4 say "epinephrine", 1 says "diphenhydramine"
    specs = [(Stub("epinephrine"), "m1"), (Stub("epinephrine"), "m2"),
             (Stub("epinephrine"), "m3"), (Stub("diphenhydramine"), "m4"),
             (Stub("epinephrine"), "m5")]
    sw = ModelSwarm(specs)
    r = sw.coordinate("First drug in anaphylaxis?", ["epinephrine", "diphenhydramine"], rounds=2)
    print("[stub] scores:", r["scores"], "decision:", r["decision"], "quorum:", r["quorum_met"])
    ok = (r["decision"] == "epinephrine" and r["quorum_met"]
          and r["scores"]["epinephrine"] > r["scores"]["diphenhydramine"])

    # abstention: a 1-1 split with quorum_min=2 should NOT force a call
    sw2 = ModelSwarm([(Stub("a"), "m1"), (Stub("b"), "m2")])
    r2 = sw2.coordinate("?", ["a", "b"], rounds=1, quorum_min=2)
    print("[stub] split scores:", r2["scores"], "decision:", r2["decision"])
    ok &= (r2["decision"] is None)        # no quorum -> abstain
    return ok


if __name__ == "__main__":
    import sys
    ok = _stub_test()
    print("SWARM STUB SELF-TEST:", "PASS" if ok else "FAIL")
    sys.exit(0 if ok else 1)

File: apple_native/JARVISNativeRuntime/swarm/swarm.cpp
#include "swarm.h"

#include <algorithm>
#include <cctype>
#include <cmath>
#include <stdexcept>
#include <unordered_set>

namespace jarvis {

static std::string lower_copy(const std::string& s) {
    std::string out;
    out.reserve(s.size());
    for (unsigned char c : s) out.push_back(static_cast<char>(std::tolower(c)));
    return out;
}

static std::string strip_chars(std::string s, const std::string& chars) {
    const auto first = s.find_first_not_of(chars);
    if (first == std::string::npos) return "";
    const auto last = s.find_last_not_of(chars);
    return s.substr(first, last - first + 1);
}

static bool contains_any_token(const std::string& text,
                               const std::vector<std::string>& tokens) {
    const std::string t = lower_copy(text);
    for (const auto& token : tokens) {
        if (t.find(token) != std::string::npos) return true;
    }
    return false;
}

static bool is_high_risk_action(const std::string& prompt, const std::string& option) {
    const std::string combined = lower_copy(prompt + " " + option);
    if (combined.find("dangerous-action") != std::string::npos) return true;
    if (combined.find("actuate") != std::string::npos) return true;
    if (combined.find("lethal") != std::string::npos) return true;
    if (combined.find("weapon") != std::string::npos) return true;
    if (combined.find("conscious") != std::string::npos &&
        (combined.find("defibrillat") != std::string::npos || combined.find("shock") != std::string::npos)) {
        return true;
    }
    if (contains_any_token(combined, {"delete", "wipe", "disable", "override", "bypass", "unlock"}) &&
        contains_any_token(combined, {"root", "credential", "operator", "production"})) {
        return true;
    }
    return false;
}

std::string swarm_norm(const std::string& label) {
    std::string out;
    out.reserve(label.size());
    bool last_us = false;
    for (unsigned char c : label) {
        const char lc = static_cast<char>(std::tolower(c));
        if ((lc >= 'a' && lc <= 'z') || (lc >= '0' && lc <= '9')) {
            out.push_back(lc);
            last_us = false;
        } else if (!last_us) {
            out.push_back('_');
            last_us = true;
        }
    }
    while (!out.empty() && out.front() == '_') out.erase(out.begin());
    while (!out.empty() && out.back() == '_') out.pop_back();
    return out;
}

std::optional<std::string> swarm_match_option(const std::string& text,
                                              const std::vector<std::string>& options) {
    const std::string t = lower_copy(text);
    std::vector<std::string> hits;
    for (const auto& option : options) {
        if (t.find(lower_copy(option)) != std::string::npos) hits.push_back(option);
    }
    if (hits.size() == 1) return hits.front();

    std::string first;
    const auto stripped = strip_chars(t, " \t\r\n");
    const auto nl = stripped.find('\n');
    first = nl == std::string::npos ? stripped : stripped.substr(0, nl);
    for (const auto& option : options) {
        const std::string opt_lower = lower_copy(option);
        if (swarm_norm(option) == swarm_norm(first) || opt_lower == strip_chars(first, " .:-")) {
            return option;
        }
    }
    return std::nullopt;
}

ModelSwarm::ModelSwarm(std::vector<SwarmAgentSpec> specs,
                       Pheromind& field,
                       Endocrine* endocrine)
    : specs_(std::move(specs)), field_(field), endocrine_(endocrine) {
    if (specs_.empty()) throw std::invalid_argument("ModelSwarm needs at least one agent spec");
    for (const auto& spec : specs_) {
        if (!spec.backend) throw std::invalid_argument("ModelSwarm agent backend must not be null");
    }
}

SwarmChatOptions ModelSwarm::chat_options_() {
    SwarmChatOptions opts;
    if (!endocrine_) return opts;

    const double cortisol = endocrine_->level("cortisol");
    const double dopamine = endocrine_->level("dopamine");
    const double adrenaline = endocrine_->level("adrenaline");

    opts.temperature = std::clamp(0.3 + 0.25 * dopamine - 0.20 * cortisol, 0.05, 0.9);
    opts.num_predict = std::max(4, static_cast<int>(std::lround(12.0 * (1.0 - 0.50 * adrenaline))));
    return opts;
}

std::optional<std::string> ModelSwarm::leader_from_field_(
    const std::unordered_map<std::string, double>& sensed) {
    if (sensed.empty()) return std::nullopt;
    auto best = sensed.begin();
    for (auto it = std::next(sensed.begin()); it != sensed.end(); ++it) {
        if (it->second > best->second) best = it;
    }
    return best->first;
}

std::optional<std::string> ModelSwarm::ask_agent(SwarmBackend& backend,
                                                 const std::string& model,
                                                 const std::string& prompt,
                                                 const std::vector<std::string>& options,
                                                 const std::optional<std::string>& leader) {
    std::string opts;
    for (std::size_t i = 0; i < options.size(); ++i) {
        if (i) opts += " | ";
        opts += options[i];
    }

    std::string hint;
    if (leader) hint = "\nThe swarm is currently leaning toward: " + *leader + ".";

    const std::string sys_msg =
        "You are one agent in a swarm deciding a single question. Weigh the swarm's "
        "current leaning, but choose what you actually judge correct. Reply with ONLY "
        "one option label, nothing else.";
    const std::string user = "Question: " + prompt + "\nOptions: " + opts + "." +
                             hint + "\nYour one-word choice:";

    try {
        const auto out = backend.chat({{"system", sys_msg}, {"user", user}}, model, chat_options_());
        return swarm_match_option(out, options);
    } catch (...) {
        return std::nullopt;
    }
}

SwarmHeadHealth ModelSwarm::head_health() const noexcept {
    return SwarmHeadHealth{specs_.size(), !specs_.empty()};
}

SwarmResult ModelSwarm::coordinate(const std::string& prompt,
                                   const std::vector<std::string>& options,
                                   int rounds,
                                   int quorum_min) {
    if (options.empty()) throw std::invalid_argument("ModelSwarm coordinate needs at least one option");
    if (quorum_min <= 0) quorum_min = std::max(2, static_cast<int>(specs_.size() / 2) + 1);

    SwarmResult result;
    result.quorum_min = quorum_min;
    result.n_agents = static_cast<int>(specs_.size());

    const int n_rounds = std::max(1, rounds);
    result.rounds.reserve(static_cast<std::size_t>(n_rounds));

    for (int r = 0; r < n_rounds; ++r) {
        (void)r;
        auto sensed = field_.sense_all(SWARM_RECRUIT_KIND);
        auto leader = leader_from_field_(sensed);
        SwarmRoundTrace trace;
        trace.leader_at_start = leader;

        for (const auto& spec : specs_) {
            auto pick = ask_agent(*spec.backend, spec.model, prompt, options, leader);
            trace.picks[spec.model] = pick;
            if (pick) {
                field_.deposit(SWARM_RECRUIT_KIND, swarm_norm(*pick), 0.34, spec.model);
            }
        }
        result.rounds.push_back(std::move(trace));
    }

    bool any_score = false;
    std::optional<std::string> decision;
    double best_score = 0.0;
    for (const auto& option : options) {
        const auto sensed = field_.sense(swarm_norm(option), std::vector<std::string>{SWARM_RECRUIT_KIND});
        const auto it = sensed.find(SWARM_RECRUIT_KIND);
        const double raw_score = it == sensed.end() ? 0.0 : it->second;
        result.scores[option] = std::round(raw_score * 10000.0) / 10000.0;
        if (raw_score > 0.0) any_score = true;
        if (!decision || raw_score > best_score) {
            decision = option;
            best_score = raw_score;
        }
    }

    if (!any_score) decision = std::nullopt;
    result.leader_raw = decision;

    std::unordered_map<std::string, std::unordered_set<std::string>> voters_by_option;
    for (const auto& trace : result.rounds) {
        for (const auto& [model, pick] : trace.picks) {
            if (pick) voters_by_option[swarm_norm(*pick)].insert(model);
        }
    }

    result.quorum_met = false;
    if (decision) {
        const auto voters = voters_by_option.find(swarm_norm(*decision));
        result.quorum_met = voters != voters_by_option.end() &&
                            static_cast<int>(voters->second.size()) >= quorum_min;
    }

    if (result.quorum_met && decision && is_high_risk_action(prompt, *decision)) {
        result.abstained_for_safety = true;
        result.decision = std::nullopt;
    } else {
        result.decision = result.quorum_met ? decision : std::nullopt;
    }
    return result;
}

} // namespace jarvis
